app/src/main/kotlin/com/intive/picover/common/viewmodel/MVIViewModel.kt

```diff
@@ -0,0 +1,32 @@
+package com.intive.picover.common.viewmodel
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.intive.picover.common.viewmodel.event.Event
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.launch
+
+abstract class MVIViewModel<T : Event> : ViewModel() {
```

Comment:
Of course this could be done in variant ways:

- Repeat everything in every view model (sucks)
- Have a big abstract class which has everything in it
- Use composition (interfaces etc.) 

I see as it as the following:

- Event is the most important thing, without this we don’t have MVI
- If User has only SideEffects, and when they are only UI then, in my opinion, viewmodel is useless :D When there are side effects more like send tracking event to firebase, the viewmodel will be needed
- state can be needed or not - depends on the screen logic and needs

So i’ve prepared events handling in base class, and effects handling in composition way and state just in the viewmodel. Thanks to that You can see different approaches to the topic. We can of course do everything using composition or have events base class + side efects + state as a additional components, it’s worth to discuss.